ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/services/admin/ApiConfigManagementService.scala

```diff
@@ -0,0 +1,207 @@
+        .toScala(
+          writeService.submitConfiguration(
+            params.maximumRecordTime,
+            request.submissionId, /* FIXME validate, SubmissionId in #3391 */
+            newConfig
+          ))
+
+      result <- submissionResult match {
+        case SubmissionResult.Acknowledged =>
+          // Ledger acknowledged. Start polling to wait for the result to land in the index.
+          // FIXME(JM): We should wait slightly longer than maxTtl? How much longer?
+          val maxTtl = Duration.fromNanos(currentConfig.timeModel.maxTtl.toNanos)
+          val timeToLive = if (params.timeToLive < maxTtl) params.timeToLive else maxTtl
+          pollUntilPersisted(request.submissionId, pollOffset, timeToLive).flatMap {
+            case accept: domain.ConfigurationEntry.Accepted =>
+              Future.successful(SetTimeModelResponse(accept.configuration.generation))
+            case rejected: domain.ConfigurationEntry.Rejected =>
+              // TODO(JM): Reconsider error kind. E.g. emit deadline exceeded when MRT exceeds?
+              Future.failed(ErrorFactories.aborted(rejected.rejectionReason))
+          }
+        case SubmissionResult.Overloaded =>
+          Future.failed(ErrorFactories.resourceExhausted("Resource exhausted"))
+        case SubmissionResult.InternalError(reason) =>
+          Future.failed(ErrorFactories.internal(reason))
+        case SubmissionResult.NotSupported =>
+          Future.failed(
+            ErrorFactories.unimplemented("Setting of time model not supported by this ledger"))
+      }
+    } yield result
+  }
+
+  private case class SetTimeModelParameters(
+      newTimeModel: v1.TimeModel,
+      maximumRecordTime: Time.Timestamp,
+      timeToLive: FiniteDuration
+  )
+
+  private def validateParameters(
+      request: SetTimeModelRequest): Either[StatusRuntimeException, SetTimeModelParameters] = {
+    import validation.FieldValidations._
+    for {
+      pTimeModel <- requirePresence(request.newTimeModel, "new_time_model")
+      pMinTransactionLatency <- requirePresence(
+        pTimeModel.minTransactionLatency,
+        "min_transaction_latency")
+      pMaxClockSkew <- requirePresence(pTimeModel.maxClockSkew, "max_clock_skew")
+      pMaxTtl <- requirePresence(pTimeModel.maxTtl, "max_ttl")
+      newTimeModel <- v1.TimeModel(
+        minTransactionLatency = DurationConversion.fromProto(pMinTransactionLatency),
+        maxClockSkew = DurationConversion.fromProto(pMaxClockSkew),
+        maxTtl = DurationConversion.fromProto(pMaxTtl)
+      ) match {
+        case Failure(err) => Left(ErrorFactories.invalidArgument(err.toString))
+        case Success(ok) => Right(ok)
+      }
+      // FIXME(JM): The maximum record time should be constrained, probably by the current active time model?
+      pMaxRecordTime <- requirePresence(request.maximumRecordTime, "maximum_record_time")
+      mrtInstant = TimestampConversion.toInstant(pMaxRecordTime)
+      timeToLive = {
```

Comment:
We need to be a bit more generous here, it takes time to propagate the data from the ledger through the indexer to ledger api service.

Reply:
yep definitely crossed my mind, but couldn't figure out what is "more generous" so left it like this. double the ttl?

Reply:
https://github.com/digital-asset/daml/pull/3756#discussion_r354748885